src/library_manager/llext_manager.c

```diff
@@ -295,21 +287,48 @@ static int llext_manager_mod_init(struct lib_manager_mod_ctx *ctx,
 				  const struct sof_man_fw_desc *desc,
 				  const struct sof_man_module *mod_array)
 {
+	unsigned int i, n_mod;
+	size_t offs;
+
+	/* count modules */
+	for (i = 0, n_mod = 0, offs = ~0; i < desc->header.num_module_entries; i++)
+		if (mod_array[i].segment[LIB_MANAGER_TEXT].file_offset != offs) {
+			offs = mod_array[i].segment[LIB_MANAGER_TEXT].file_offset;
+			n_mod++;
+		}
+
 	/*
 	 * Loadable modules are loaded to DRAM once and never unloaded from it.
 	 * Context, related to them, is never freed
 	 */
 	ctx->mod = rmalloc(SOF_MEM_ZONE_RUNTIME_SHARED, SOF_MEM_FLAG_COHERENT,
-			   SOF_MEM_CAPS_RAM, sizeof(ctx->mod[0]));
+			   SOF_MEM_CAPS_RAM, n_mod * sizeof(ctx->mod[0]));
 	if (!ctx->mod)
 		return -ENOMEM;
 
-	ctx->n_mod = 1;
-	ctx->mod[0].start_idx = 0;
+	ctx->n_mod = n_mod;
+
+	for (i = 0, n_mod = 0, offs = ~0; i < desc->header.num_module_entries; i++)
+		if (mod_array[i].segment[LIB_MANAGER_TEXT].file_offset != offs) {
+			offs = mod_array[i].segment[LIB_MANAGER_TEXT].file_offset;
+			ctx->mod[n_mod].segment[LIB_MANAGER_TEXT].size = 0;
+			ctx->mod[n_mod++].start_idx = i;
+		}
 
 	return 0;
 }
 
+static unsigned int llext_manager_mod_find(const struct lib_manager_mod_ctx *ctx, unsigned int idx)
+{
+	unsigned int i;
+
+	for (i = 0; i < ctx->n_mod; i++)
+		if (ctx->mod[i].start_idx > idx)
+			break;
+
+	return i - 1;
```

Comment:
what if "ctx->mod[0].startidx > idx" is true? this would return (unsigned)-1

Reply:
@kv2019i it cannot be true by construction - the first module always starts from 0. I can add a comment in an incremental PR